components/script/dom/abortsignal.rs

```diff
@@ -264,33 +276,84 @@ impl AbortSignal {
                 result_signal
                     .source_signals
                     .borrow_mut()
-                    .insert(Dom::from_ref(signal));
+                    .insert(WeakRef::new(signal));
                 // Step 4.1.2. Append resultSignal to signal’s dependent signals.
                 signal
                     .dependent_signals
                     .borrow_mut()
-                    .insert(Dom::from_ref(&result_signal));
+                    .insert(WeakRef::new(&*result_signal));
             } else {
                 // Step 4.2. Otherwise, for each sourceSignal of signal’s source signals:
-                for source_signal in signal.source_signals.borrow().iter() {
-                    // Step 4.2.1. Assert: sourceSignal is not aborted and not dependent.
-                    assert!(!source_signal.aborted() && !source_signal.dependent.get());
-                    // Step 4.2.2. Append sourceSignal to resultSignal’s source signals.
-                    result_signal
-                        .source_signals
-                        .borrow_mut()
-                        .insert(source_signal.clone());
-                    // Step 4.2.3. Append resultSignal to sourceSignal’s dependent signals.
-                    source_signal
-                        .dependent_signals
-                        .borrow_mut()
-                        .insert(Dom::from_ref(&result_signal));
+                for source_signal_weak in signal.source_signals.borrow().iter() {
+                    if let Some(source_signal) = source_signal_weak.root() {
+                        // Step 4.2.1. Assert: sourceSignal is not aborted and not dependent.
+                        assert!(!source_signal.aborted() && !source_signal.dependent.get());
+                        // Step 4.2.2. Append sourceSignal to resultSignal’s source signals.
+                        result_signal
+                            .source_signals
+                            .borrow_mut()
+                            .insert(WeakRef::new(&*source_signal));
+                        // Step 4.2.3. Append resultSignal to sourceSignal’s dependent signals.
+                        source_signal
+                            .dependent_signals
+                            .borrow_mut()
+                            .insert(WeakRef::new(&*result_signal));
+                    }
+                    // else: underlying source signal was GC'd; we can just ignore this entry.
                 }
             }
         }
         // Step 5. Return resultSignal.
+        global.register_dependent_abort_signal(&result_signal);
         result_signal
     }
+
+    /// Remove weak references whose target has been collected.
+    fn prune_dead_weak_refs(&self) {
```

Comment:
Is this necessary? Note that this will run only on the dependent signals that have been registered with the global. 
I would have assumed just letting the js refs be gc'ed would be enough.

Reply:
If I understand correctly then this is not relevant for correctness, it just frees up the memory for pointers to signals that no longer exist.

Reply:
indeed this is just to pointers to signals that no longer exist, but I can remove it

Reply:
Thanks for the info. Seems fine to keep it to me.